File: 02_automatedSimplifiedExtensionLayer/00_matlab/functions/write2model.m
function Model =  write2model(process_list)

%% create empty sheets
technical_flows = [];
A_distribution = [];
A_mean = [];
A_std_dev = [];
factor_requirements = [];
F_distribution = [];
F_mean = [];
F_std_dev = [];

for i = 1:length(process_list)
   
    %% Add Process Info
    processes{1,i} = process_list(i).info.name;
    processes{2,i} = process_list(i).info.abbrevation;
    processes{3,i} = process_list(i).info.process_description;
    processes{4,i} = process_list(i).info.mainflow;
    processes{5,i} = process_list(i).info.location;
    processes{6,i} = process_list(i).info.exact_location;
    processes{7,i} = process_list(i).info.capacity;
    processes{8,i} = process_list(i).info.unit_per_year;
    processes{9,i} = '';
    processes{10,i} = '';
    
    %% Technical Flows 
    % search if stream that already exists 
    for k = 1:length(process_list(i).streams)
        flag = 1;
        for j = 1:size(technical_flows,1) 
            if strcmp(technical_flows{j,1},process_list(i).streams(k).name) && (technical_flows{j,2} ==  process_list(i).streams(k).class || k==1) && strcmp(technical_flows{j,6},process_list(i).streams(k).amount_unit)
                
                A_distribution{j,i} = 2;
                A_mean{j,i} = process_list(i).streams(k).amount;
                
                if isnan(A_mean{j,i})
                    A_mean{j,i} = process_list(i).streams(k).cost_per_kg;
                end
                
                A_std_dev{j,i} = 0;
                flag = 0;
                break;
            end
        end
        
        %% add new flow to tec_streams
        if flag
            technical_flows{end+1,1} =  process_list(i).streams(k).name; %name
            technical_flows{end,2} =  process_list(i).streams(k).class;  %category
            technical_flows{end,3} = ''; %concentration
            technical_flows{end,4} = ''; %CAS
            technical_flows{end,5} = process_list(i).streams(k).unit_type; %unit typ
            if isnan(process_list(i).streams(k).cost_unit)
                technical_flows{end,5} = 'Pieces';
            end
            technical_flows{end,6} = process_list(i).streams(k).amount_unit; %unit
            if isnan(process_list(i).streams(k).cost_unit)
                technical_flows{end,6} = '$';
            end
            technical_flows{end,7} = process_list(i).streams(k).cost; %cost
            if isnan(process_list(i).streams(k).cost)
                technical_flows{end,7} = 1;
            end
            technical_flows{end,8} = 0; %LHV
            technical_flows{end,9} = 0; %HHV
            technical_flows{end,10} = 0; %Formula 
            technical_flows{end,11} = '';
            technical_flows{end,12} = '';
            technical_flows{end,13} = '';
            technical_flows{end,14} = '';
            technical_flows{end,15} = '';
            technical_flows{end,16} = '';
            technical_flows{end,17} = '';
            
            if isempty(j)
                j = 0;
            end
            A_distribution{j+1,i} = 2;
            A_mean{j+1,i} = process_list(i).streams(k).amount;
            A_std_dev{j+1,i} = 0;
            if isnan(A_mean{j+1,i})
                A_mean{j+1,i} = process_list(i).streams(k).cost_per_kg;
            end
        end
    end
    
    %% Factor requirement flows
    for k = 1:length(process_list(i).costs)
        flag = 1;
        for j = 1:size(factor_requirements,1) 
            if strcmp(factor_requirements{j,1},process_list(i).costs(k).name)
                F_distribution{j,i} = 2;
                F_mean{j,i} = process_list(i).costs(k).value;
                F_std_dev{j,i} = 0;
                flag = 0;
                break;
            end
        end    
        %% add new flow to tec_streams
        if flag
            factor_requirements{end+1,1} =  process_list(i).costs(k).name; %name
            factor_requirements{end,2} = process_list(i).costs(k).unit;
            factor_requirements{end,3} = '';
            
            if isempty(j)
                j = 0;
            end
            
            F_distribution{j+1,i} = 2;
            F_mean{j+1,i} = process_list(i).costs(k).value;
            F_std_dev{j+1,i} = 0;
        end
    end
end

%% create k 

k_mean = cell(size(F_mean,1),1);
k_distribution = cell(size(F_mean,1),1);
k_std_dev = cell(size(F_mean,1),1);

k_mean(:,1) = {1}; 
k_distribution(1:end) = {2}; 


%% Add legend to Tables

%% Technical Flows
tech_name{1,1} = 'name';
tech_name{1,2} = 'category';
tech_name{1,3} = 'concentration/purity';
tech_name{1,4} = 'CAS-Nr.';
tech_name{1,5} = 'unit(choice)';
tech_name{1,6} = '[unit choice]';
tech_name{1,7} = 'Market price';
tech_name{1,8} = 'LHV';
tech_name{1,9} = 'HHV';
tech_name{1,10} = 'chemical formular[C2C4 format]';
tech_name{1,11} = 'location(choice)';
tech_name{1,12} = 'exact location';
tech_name{1,13} = 'comments';
tech_name{1,14} = 'SMILES CODE';
tech_name{1,15} = 'molecular mass';
tech_name{1,16} = 'flowCategory';
tech_name{1,17} = 'flowSubcategory';

%% Processes
process_names{1,1} = 'name';
process_names{2,1} = 'abbrevation';
process_names{3,1} = 'process description';
process_names{4,1} = 'mainflow';
process_names{5,1} = 'location (choice)';
process_names{6,1} = 'exact location';
process_names{7,1} = 'capacity';
process_names{8,1} = 'unit per year (choice)';
process_names{9,1} = 'comments';
process_names{10,1} = 'type';
%% factor requierments
req_names{1,1} = 'name';
req_names{1,2} = 'unit';
req_names{1,3} = 'comment';

%% cost
cost_name{1,1} = 'name';
cost_name{1,2} = 'cost';

%% Add the legends 
processes = horzcat(process_names,processes);
technical_flows = vertcat(tech_name,technical_flows);
factor_requirements = vertcat(req_names,factor_requirements);

% process_names = processes(1,1:end);
% flow_names = technical_flows(2:end,1);
% requirement_names = factor_requirements(2:end,1);

% A_distribution = horzcat(flow_names,A_distribution);
% A_distribution = vertcat(process_names,A_distribution);

% A_mean = horzcat(flow_names,A_mean);
% A_mean = vertcat(process_names,A_mean);

% A_std_dev = horzcat(flow_names,A_std_dev);
% A_std_dev = vertcat(process_names,A_std_dev);

% k_mean = horzcat(requirement_names,k_mean);
% k_mean = vertcat(cost_name,k_mean);
% 
% k_std_dev = horzcat(requirement_names,k_std_dev);
% k_std_dev = vertcat(cost_name,k_std_dev);
% 
% k_distribution = horzcat(requirement_names,k_distribution);
% k_distribution = vertcat(cost_name,k_distribution);
% 
% F_distribution = horzcat(requirement_names,F_distribution);
% F_distribution = vertcat(process_names,F_distribution);
% 
% F_mean = horzcat(requirement_names,F_mean);
% F_mean = vertcat(process_names,F_mean);
% 
% F_std_dev = horzcat(requirement_names,F_std_dev);
% F_std_dev = vertcat(process_names,F_std_dev);


Model.meta_data_flows = technical_flows;
Model.meta_data_processes = processes;

A_mean(find(cellfun(@(C)...
any(isempty(C(:))), A_mean)))={0};  

A_mean = cell2mat(A_mean);

Model.matrices.A.mean_values = A_mean;

% Model.matrices.A.distribution = A_distribution;
% Model.matrices.A.std_dev = A_std_dev;

Model.meta_data_factor_requirements = factor_requirements;

F_mean(find(cellfun(@(C)...
any(isempty(C(:))), F_mean)))={0};  

F_mean = cell2mat(F_mean);

Model.matrices.F.mean_values = F_mean;

% Model.matrices.F.distribution = F_distribution;
% Model.matrices.F.std_dev = F_std_dev;

k_mean(find(cellfun(@(C)...
any(isempty(C(:))), k_mean)))={0};  

k_mean = cell2mat(k_mean);

Model.matrices.k.mean_values = k_mean;

% Model.matrices.k.distribution = k_distribution;
% Model.matrices.k.std_dev = k_std_dev;

end

File: 02_automatedSimplifiedExtensionLayer/utils/write2model.py
import numpy as np

def write2model(process_list):
    Model = {}

    # Create empty sheets
    technical_flows = []
    A_distribution = []
    A_mean = []
    A_std_dev = []
    factor_requirements = []
    F_distribution = []
    F_mean = []
    F_std_dev = []

    processes = []

    for i in range(len(process_list)):
        # Add Process Info
        process_info = process_list[i].info
        processes.append([
            process_info.name,
            process_info.abbrevation,
            process_info.process_description,
            process_info.mainflow,
            process_info.location,
            process_info.exact_location,
            process_info.capacity,
            process_info.unit_per_year,
            '', ''  # Comments and type placeholders
        ])

        # Technical Flows
        for k in range(len(process_list[i].streams)):
            flag = 1
            for j in range(len(technical_flows)):
                stream = process_list[i].streams[k]
                if (technical_flows[j][0] == stream.name and 
                    technical_flows[j][1] == stream.stream_class or k == 1 and 
                    technical_flows[j][5] == stream.amount_unit):

                    # Ensure lists are large enough
                    if len(A_distribution) <= j:
                        A_distribution.append([None] * len(process_list))
                        A_mean.append([None] * len(process_list))
                        A_std_dev.append([None] * len(process_list))

                    A_distribution[j][i] = 2
                    A_mean[j][i] = stream.amount if not np.isnan(stream.amount) else stream.cost_per_kg
                    A_std_dev[j][i] = 0
                    flag = 0
                    break

            # Add new flow to technical_flows if not found
            if flag:
                stream = process_list[i].streams[k]
                technical_flow_entry = [
                    stream.name, 
                    stream.stream_class, 
                    '', '',  # Concentration, CAS
                    stream.unit_type if not np.isnan(stream.cost_unit) else 'Pieces',
                    stream.amount_unit if not np.isnan(stream.cost_unit) else '$',
                    stream.cost if not np.isnan(stream.cost) else 1,
                    0, 0,  # LHV, HHV
                    0, '', '', '', '', '', ''  # Additional fields
                ]
                technical_flows.append(technical_flow_entry)

                # Update distributions and mean values
                j = len(technical_flows) - 1
                A_distribution.append([None] * len(process_list))
                A_mean.append([None] * len(process_list))
                A_std_dev.append([None] * len(process_list))

                A_distribution[j][i] = 2
                A_mean[j][i] = stream.amount if not np.isnan(stream.amount) else stream.cost_per_kg
                A_std_dev[j][i] = 0

        # Factor requirement flows
        for k in range(len(process_list[i].costs)):
            flag = 1
            for j in range(len(factor_requirements)):
                cost = process_list[i].costs[k]
                if factor_requirements[j][0] == cost.name:
                    if len(F_distribution) <= j:
                        F_distribution.append([None] * len(process_list))
                        F_mean.append([None] * len(process_list))
                        F_std_dev.append([None] * len(process_list))

                    F_distribution[j][i] = 2
                    F_mean[j][i] = cost.value
                    F_std_dev[j][i] = 0
                    flag = 0
                    break

            # Add new flow to factor_requirements if not found
            if flag:
                cost = process_list[i].costs[k]
                factor_requirement_entry = [cost.name, cost.unit, '']
                factor_requirements.append(factor_requirement_entry)

                j = len(factor_requirements) - 1
                F_distribution.append([None] * len(process_list))
                F_mean.append([None] * len(process_list))
                F_std_dev.append([None] * len(process_list))

                F_distribution[j][i] = 2
                F_mean[j][i] = cost.value
                F_std_dev[j][i] = 0

    # Create k matrix
    k_mean = [[1] for _ in range(len(F_mean))]
    k_distribution = [[2] for _ in range(len(F_mean))]
    k_std_dev = [[0] for _ in range(len(F_mean))]

    # Add legends
    tech_name = [
        ['name', 'category', 'concentration/purity', 'CAS-Nr.', 'unit(choice)', 
         '[unit choice]', 'Market price', 'LHV', 'HHV', 'chemical formula', 
         'location(choice)', 'exact location', 'comments', 'SMILES CODE', 
         'molecular mass', 'flowCategory', 'flowSubcategory']
    ]
    
    process_names = [
        ['name', 'abbrevation', 'process description', 'mainflow', 
         'location (choice)', 'exact location', 'capacity', 'unit per year (choice)', 
         'comments', 'type']
    ]
    
    req_names = [['name', 'unit', 'comment']]
    cost_name = [['name', 'cost']]  # Defined cost_name to be used with cost-related data

    # Concatenate legends with actual data
    processes = process_names + processes
    technical_flows = tech_name + technical_flows
    factor_requirements = req_names + factor_requirements

    # Assign matrices to Model structure
    Model['meta_data_flows'] = technical_flows
    Model['meta_data_processes'] = processes

    # Handle empty cells in A_mean and convert to numpy array
    A_mean = [[0 if item is None else item for item in row] for row in A_mean]
    A_mean = np.array(A_mean)
    Model['matrices'] = {'A': {'mean_values': A_mean}}

    # Same for F_mean
    F_mean = [[0 if item is None else item for item in row] for row in F_mean]
    F_mean = np.array(F_mean)
    Model['matrices']['F'] = {'mean_values': F_mean}

    # Same for k_mean, k_distribution, k_std_dev
    k_mean = [[0 if item is None else item for item in row] for row in k_mean]
    k_mean = np.array(k_mean)
    Model['matrices']['k'] = {
        'mean_values': k_mean,
        'distribution': np.array(k_distribution),
        'std_dev': np.array(k_std_dev)
    }

    # Use cost_name to create a table or include it in the Model
    Model['meta_data_costs'] = cost_name  # Added cost_name as part of Model

    return Model
